File: src/options.js
"use strict";
exports=module.exports;

var Constants = require('./constants');
var Constants = require('./constants.js');
var CompoundTerm = require('./compound_term.js');
var AtomTerm = require('./atom_term.js');
var VariableTerm = require('./variable_term.js');
var IntegerTerm = require('./integer_term.js');
var CTable = require('./ctable.js');

module.exports.parseOptions = function(t, domain)
{
    var list = t;
    var options = {};
    while (TAGOF(list) == CompoundTag && FUNCTOROF(list) == Constants.listFunctor)
    {
        var head = ARGOF(list, 0);
        list = ARGOF(list, 1);
        if (TAGOF(head) == VariableTag)
            Errors.instantiationError(head);
        // FIXME: In many cases we are supposed to check that head is in some domain
        if (TAGOF(head) == ConstantTag && CTable.get(head) instanceof AtomTerm)
            options[CTable.get(head).value] = true;
        else if (TAGOF(head) == CompoundTag)
        {
            var functor = CTable.get(FUNCTOROF(head));
            if (functor.arity == 1)
            {
                var h0 = ARGOF(head,0);
                if (TAGOF(h0) == ConstantTag && CTable.get(h0) instanceof AtomTerm)
                {
                    h0 = CTable.get(h0);
                    if (h0.value == "true")
                        options[CTable.get(functor.name).value] = true;
                    else if (h0.value == "false")
                        options[CTable.get(functor.name).value] = false;
                    else
                        options[CTable.get(functor.name).value] = h0.value;
                }
                else if (TAGOF(h0) == ConstantTag && CTable.get(h0) instanceof IntegerTerm)
                    options[CTable.get(functor.name).value] = CTable.get(h0).value;
                Errors.domainError(domain, head);
            }
            Errors.domainError(domain, head);
        }
        else
            Errors.domainError(domain, head);
    }
    if (TAGOF(list) == VariableTag)
        Errors.instantiationError(list);
    if (list != Constants.emptyListAtom)
        Errors.typeError(Constants.listAtom, t);
    return options;
}

File: src/options.c
#include "options.h"
#include "constants.h"
#include "errors.h"
#include "ctable.h"
#include "whashmap.h"
#include "kernel.h"

void init_options(Options* o)
{
   o->map = whashmap_new();
}
void free_options(Options* o)
{
   whashmap_free(o->map);
}

void set_option(Options* o, word key, word value)
{
   whashmap_put(o->map, key, (any_t)value);
}

int options_from_term(Options* o, word t)
{
   while(TAGOF(t) == COMPOUND_TAG && FUNCTOROF(t) == listFunctor)
   {
      word head = ARGOF(t, 0);
      if (TAGOF(head) != COMPOUND_TAG)
         return type_error(optionAtom, head);
      Functor f = getConstant(FUNCTOROF(head)).data.functor_data;
      if (f->arity > 2)
         return type_error(optionAtom, head);
      if (f->arity == 1)
         set_option(o, f->name, trueAtom);
      else
         set_option(o, f->name, ARGOF(t, 0));
      t = ARGOF(t, 1);
   }
   return t == emptyListAtom;
}

word get_option(Options* o, word key, word defaultvalue)
{
   word w;
   if (whashmap_get(o->map, key, (any_t*)&w) == MAP_OK)
      return w;
   return defaultvalue;
}
